### PR TITLE
jv-fruit-shop

### PR DESCRIPTION
**Pull request by alwayswannajava**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/DataConverterImpl.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    private static final int FRUIT_OPERATION_INDEX = 0;
+    private static final int FRUIT_NAME_INDEX = 1;
+    private static final int FRUIT_QUANTITY_INDEX = 2;
+
+    @Override
+    public List<Fruit> convertToFruit(List<String> inputReport) {
+        return inputReport.stream()
+                .skip(COUNT_HEADER_LINES_TO_SKIP)
+                .map(line -> line.split(COMMA_DELIMITER))
+                .map(line -> new Fruit(
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/DataConverterImpl.java:19]: 2 methods map not best practice

File: src/main/java/core/basesyntax/service/ReportServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+public class ReportServiceImpl implements ReportService {
+    private static final String HEADER_LINE = "fruit, quantity";
+    private static final String COMMA_DELIMITER = ",";
+
+    @Override
+    public String generateReport() {
+        StringBuilder reportBuilder = new StringBuilder();
+        reportBuilder.append(HEADER_LINE)
+                .append("\n");
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/ReportServiceImpl.java:13]: better use lineSeparator

File: src/main/java/core/basesyntax/service/ShopServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.domain.Fruit;
+import core.basesyntax.service.strategy.OperationStrategy;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private static final String APPLE_NAME = "apple";
+    private static final String BANANA_NAME = "banana";
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/ShopServiceImpl.java:10]: maybe better use enum for this

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -11,10 +11,10 @@ Your tasks are:
...
 There are four activities at the store:
 ```text
-    b - balance, the remnants of fruits at the beginning of the working day
-    s - supply, means you are receiving new fruits from suppliers
-    p - purchase, means someone has bought some fruit
-    r - return, means someone who has bought the fruits now returns them back
+    b - balance, the remnants of fruitTransactions at the beginning of the working day
+    s - supply, means you are receiving new fruitTransactions from suppliers
+    p - purchase, means someone has bought some fruitTransaction
+    r - return, means someone who has bought the fruitTransactions now returns them back
```
Comment (MentorMateAcademy) [inline README.md:17]: undo this changes

File: pom.xml [hunk 5]
```diff
@@ -5,7 +5,7 @@
     <modelVersion>4.0.0</modelVersion>
 
     <groupId>mate-academy</groupId>
-    <artifactId>jv-fruit-shop</artifactId>
+    <artifactId>jv-fruitTransaction-shop</artifactId>
```
Comment (MentorMateAcademy) [inline pom.xml:8]: don`t need change this

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -70,7 +70,7 @@ The line above means you have 152 bananas, and 90 apples in your Fruit store aft
 ```java
 public class FruitTransaction {
   private Operation operation;
-  private String fruit;
+  private String fruitTransaction;
```
Comment (MentorMateAcademy) [inline README.md:73]: Undo this changes in readme and checklist files

File: src/main/java/core/basesyntax/FruitShopApplication.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        List<String> readLines = fileService.read("fruits.csv");
+        DataConverterService dataConverterService = new DataConverterImpl();
+        final List<FruitTransaction> convertedFruitTransactions =
+                dataConverterService.convertToFruit(readLines);
+        Map<FruitTransaction.Operation, OperationHandler> operationHandlers = new HashMap<>();
+        operationHandlers.put(FruitTransaction.Operation.BALANCE, new BalanceOperation());
+        operationHandlers.put(FruitTransaction.Operation.PURCHASE, new PurchaseOperation());
+        operationHandlers.put(FruitTransaction.Operation.RETURN, new ReturnOperation());
+        operationHandlers.put(FruitTransaction.Operation.SUPPLY, new SupplyOperation());
+        OperationStrategy operationStrategy = new OperationStrategyImpl(operationHandlers);
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/FruitShopApplication.java:36]: use better map.of

File: src/main/java/core/basesyntax/service/DataConverterImpl.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    private FruitTransaction convertToFruitTransaction(String line) {
+        String[] splitFruits = line.split(COMMA_DELIMITER);
+        FruitTransaction.Operation fruitOperation =
+                FruitTransaction.Operation.getByCode(splitFruits[FRUIT_OPERATION_INDEX]);
+        FruitTransaction.FruitName fruitName =
+                FruitTransaction.FruitName.getByFruitName(splitFruits[FRUIT_NAME_INDEX]);
+        int fruitQuantity = Integer.parseInt(splitFruits[FRUIT_QUANTITY_INDEX]);
+        return new FruitTransaction(fruitOperation, fruitName, fruitQuantity);
+    }
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/DataConverterImpl.java:30]: This approach looks so good)

File: src/main/java/core/basesyntax/service/DataConverterImpl.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                .map(this::convertToFruitTransaction)
+                .collect(Collectors.toList());
+    }
+
+    private FruitTransaction convertToFruitTransaction(String line) {
+        String[] splitFruits = line.split(COMMA_DELIMITER);
+        FruitTransaction.Operation fruitOperation =
+                FruitTransaction.Operation.getByCode(splitFruits[FRUIT_OPERATION_INDEX]);
+        FruitTransaction.FruitName fruitName =
+                FruitTransaction.FruitName.getByFruitName(splitFruits[FRUIT_NAME_INDEX]);
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/DataConverterImpl.java:27]:
```suggestion
        FruitTransaction.Operation fruitOperation =
                FruitTransaction.Operation.getByCode(splitFruits[FRUIT_OPERATION_INDEX]);
                
        FruitTransaction.FruitName fruitName =
                FruitTransaction.FruitName.getByFruitName(splitFruits[FRUIT_NAME_INDEX]);
```

Comment (MentorMateAcademy) [reply src/main/java/core/basesyntax/service/DataConverterImpl.java:27]: divide this by empty line, cause in github it looks so strange

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: checklist.md [hunk 10]
```diff
@@ -1,21 +1,21 @@
 ## Common mistakes (jv-fruit-shop)
 
-#### Don't begin class or method implementation with empty line. 
+#### Don't begin class or method implementation with empty line.
```
Comment (okuzan) [inline checklist.md:3]: undo changes to this file

File: src/main/java/core/basesyntax/domain/FruitTransaction.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+        public static FruitName getByFruitName(String name) {
+            return Arrays.stream(FruitName.values())
+                    .filter(fruitName -> fruitName.name.equals(name))
+                    .findFirst()
+                    .orElseThrow(() -> new RuntimeException("Invalid fruit name: " + name));
+        }
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/domain/FruitTransaction.java:91]:
```suggestion
```

File: checklist.md [hunk 12]
```diff
@@ -30,32 +30,32 @@ Remember about the informative commit and PR naming. Person that is outside of c
 what feature/functional you have added.
 
 #### Don't ignore exceptions in try-catch blocks
-Don't leave `e.printStackTrace()` or `System.out.println(e.getMessage())` in the catch blocks. 
+Don't leave `e.printStackTrace()` or `System.out.println(e.getMessage())` in the catch blocks.
 Let's rethrow a RuntimeException with an **informative** message and exception object.
 
-- Good:   
+- Good:
```
Comment (okuzan) [inline checklist.md:36]: please undo all unintended changes to this file

File: src/main/java/core/basesyntax/domain/FruitTransaction.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        FruitName(String name) {
+            this.name = name;
+        }
+
+        public static FruitName getByFruitName(String name) {
+            return Arrays.stream(FruitName.values())
+                    .filter(fruitName -> fruitName.name.equals(name))
+                    .findFirst()
+                    .orElseThrow(() -> new RuntimeException("Invalid fruit name: " + name));
```
Comment (okuzan) [inline src/main/java/core/basesyntax/domain/FruitTransaction.java:87]: let's throw more specific nosuchelementexception

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/operation/PurchaseOperation.java [hunk 14]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service.operation;
+
+public class PurchaseOperation implements OperationHandler {
+    @Override
+    public int getQuantity(int defaultQuantity, int fruitQuantity) {
+        return defaultQuantity - fruitQuantity;
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/operation/PurchaseOperation.java:6]: Do you validate that you can`t reach negative value?

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/operation/PurchaseOperation.java [hunk 15]
```diff
@@ -3,6 +3,14 @@
 public class PurchaseOperation implements OperationHandler {
     @Override
     public int getQuantity(int defaultQuantity, int fruitQuantity) {
+        validateQuantity(defaultQuantity, fruitQuantity);
         return defaultQuantity - fruitQuantity;
     }
+
+    private void validateQuantity(int defaultQuantity, int fruitQuantity) {
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/service/operation/PurchaseOperation.java:10]: why did you validate the result of the operation instead of provided value?

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]
please fix changes in checklist to default

File: src/main/java/core/basesyntax/service/ShopServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+            if (currentFruitTransaction.getName() == FruitTransaction.FruitName.APPLE) {
+                appleQuantity = operationStrategy.get(currentOperation)
+                        .getQuantity(appleQuantity, currentQuantity);
+            } else if (currentFruitTransaction.getName() == FruitTransaction.FruitName.BANANA) {
+                bananaQuantity = operationStrategy.get(currentOperation)
+                        .getQuantity(bananaQuantity, currentQuantity);
+            }
+        }
+        fruitDao.add(new FruitTransaction(FruitTransaction.FruitName.APPLE, appleQuantity));
+        fruitDao.add(new FruitTransaction(FruitTransaction.FruitName.BANANA, bananaQuantity));
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/ShopServiceImpl.java:33]: what if will be 10 or 100 kinds of fruit? it looks like duplication of code, make this method more combine and universal

File: src/main/java/core/basesyntax/dao/FruitDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.domain.FruitTransaction;
+
+public class FruitDaoImpl implements FruitDao {
+    @Override
+    public void add(FruitTransaction fruitTransaction) {
+        Storage.getFruits().add(fruitTransaction);
+    }
+}
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/dao/FruitDaoImpl.java:11]: is make sense for one method, maybe better write it in storage or write here all crud operations here

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Review by okuzan — APPROVED** [review]
